src/Generator/EnumGenerator/Cases/CaseFactory.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace Laminas\Code\Generator\EnumGenerator\Cases;
+
+use Exception;
+use Laminas\Code\Generator\EnumGenerator\CasesInterface;
+
+use function array_key_exists;
+
+final class CaseFactory
+{
+    /**
+     * @psalm-param array{
+     *      name: string,
+     *      pureCases?: list<string>,
+     *      backedCases?: array{
+     *          type: string,
+     *          cases: array<string, int|string>
+     *      },
+     *      methods?: list<array{
+     *          name: string,
+     *          visibility?: string,
+     *          static?: bool,
+     *          returntype?: string,
+     *      }>
+     * } $options
+     */
+    public static function fromOptions(array $options): CasesInterface
+    {
+        $isPureCasesSpecified   = array_key_exists('pureCases', $options);
+        $isBackedCasesSpecified = array_key_exists('backedCases', $options);
+
+        if ($isPureCasesSpecified && $isBackedCasesSpecified === false) {
+            return self::fromPureCases($options['pureCases'] ?? []);
+        }
+
+        if ($isPureCasesSpecified === false && $isBackedCasesSpecified) {
+            return self::fromBackedCases($options['backedCases'] ?? ['type' => '', 'cases' => []]);
+        }
+
+        /** @todo add more specific exception */
```

Comment:
Yes, I'll add a new exception. Do you have any suggestions what could we name it? `InvalidCasesSpecificationException` maybe?

Reply:
See comments around this: if we refine the types, the exception is not needed

Reply:
I tried to remove the exception thrown, but no luck. since it's possible for people in userland wrongly send options without neither property set, it seems we need to throw the exception when it happens. Do you see another way around this?

Reply:
never mind, using the `assert` above the return of the `backedCases` worked.